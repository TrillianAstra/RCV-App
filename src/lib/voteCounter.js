//@flow
import type { Vote, Results, Round, Totals, CandidateId } from './voteTypes';

export const getResults = (
  votes: Array<any>,
  candidates: Array<CandidateId>
): Results => {
  const losers: Array<CandidateId> = [];
  const results: Results = { rounds: [], winner: '' };
  do {
    const round = countRound(candidates, votes, losers);
    const loser = round.loser;
    round.previousLosers = losers.slice();
    results.rounds.push(round);
    if (loser) losers.push(loser);
  } while (!results.rounds.slice(-1)[0].winner);

  results.winner = results.rounds.slice(-1)[0].winner;

  return results;
};

const remainingCandidates = (
  candidates: Array<CandidateId>,
  losers: Array<CandidateId>
) => candidates.filter(c => !losers.includes(c));

const countRound = (
  candidates: Array<CandidateId>,
  votes: Array<Vote>,
  losers: Array<CandidateId>
): Round => {
  const remainder = remainingCandidates(candidates, losers);
  const round = votes.reduce((round: Round, vote: Vote) => {
    const candidate = resolveVote(vote, remainder);
    if (candidate) {
      const favorite = vote[1];
      const previousCount = round.segments[candidate].get(favorite) || 0;
      round.segments[candidate].set(favorite, previousCount + 1);
      const total = round.totals[candidate] || 0;
      round.totals[candidate] = total + 1;
      round.validVoteCount = round.validVoteCount + 1;
    }
    return round;
  }, emptyRound(candidates, losers));
  round.winner = winner(round, candidates);
  round.loser = loser(round.totals, remainder);
  round.previousLosers = losers;
  return round;
};

const emptyRound = (
  candidates: Array<CandidateId>,
  losers: Array<CandidateId>
): Round => {
  const round = {
    segments: {},
    totals: {},
    validVoteCount: 0,
    loser: null,
    winner: null,
    previousLosers: []
  };
  const loserVotes = losers.map(loser => [loser, 0]);
  const remaining = remainingCandidates(candidates, losers).map(c => [c, 0]);
  candidates.forEach(c => {
    round.totals[c] = 0;
    const newSegmentMap = new Map([
      [c, 0],
      ...loserVotes,
      ...remaining.filter(s => s[0] !== c)
    ]);
    round.segments[c] = newSegmentMap; //new Map([[c, 0], ...competitorVotes, ...loserVotes]);
  });
  return round;
};

const resolveVote = (
  vote: Vote,
  candidates: Array<CandidateId>
): ?CandidateId => {
  let position = 1;
  //For now we're bailing on skipped-rank votes
  while (vote[position]) {
    //find first candidate still in the running
    if (candidates.includes(vote[position])) return vote[position];
    position = position + 1;
  }
  return null;
};

const winner = (round: Round, candidates: Array<CandidateId>): ?CandidateId => {
  const goal = round.validVoteCount / 2;
  return candidates.find(candidate => {
    const count = round.totals[candidate];
    return count > goal;
  });
};

const loser = (totals: Totals, candidates: Array<CandidateId>): CandidateId => {
  let loser = '';
  let min: number;
  candidates.forEach((candidate: CandidateId) => {
    const count = totals[candidate];
    if (!min || count < min) {
      loser = candidate;
      min = count;
    }
  });
  return loser;
};
