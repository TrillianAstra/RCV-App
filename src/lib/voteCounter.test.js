import {getResults} from './voteCounter';

const earlyWin = [
  {1: 'D', 2: 'B', 3: 'C', 4: 'A', 5: 'E'},
  {1: 'D', 2: 'C', 3: 'B', 4: 'A', 5: 'E'},
  {1: 'D', 2: 'E', 3: 'C', 4: 'A', 5: 'B'},
  {1: 'C', 2: 'B', 3: 'A', 4: 'D', 5: 'E'},
  {1: 'E', 2: 'B', 3: 'C', 4: 'D', 5: 'A'},
  {1: 'A', 2: 'D', 3: 'C', 4: 'B', 5: 'E'},
  {1: 'D', 2: 'A', 3: 'C', 4: 'E', 5: 'B'},
];

const balanced = [
  {1: 'B', 2: 'D', 3: 'C', 4: 'A', 5: 'E'},
  {1: 'D', 2: 'C', 3: 'B', 4: 'A', 5: 'E'},
  {1: 'E', 2: 'A', 3: 'C', 4: 'D', 5: 'B'},
  {1: 'C', 2: 'B', 3: 'A', 4: 'D', 5: 'E'},
  {1: 'E', 2: 'B', 3: 'C', 4: 'D', 5: 'A'},
  {1: 'A', 2: 'D', 3: 'C', 4: 'B', 5: 'E'},
  {1: 'D', 2: 'A', 3: 'C', 4: 'E', 5: 'B'},
];

const candidates = ['A', 'B', 'C', 'D', 'E'];

it('should run only round 0 and return results', () => {
  const {rounds} = getResults(earlyWin, candidates);

  expect(rounds.length).toEqual(1);
  expect(rounds[0].winners[0]).toEqual('D');
  expect(rounds).toMatchSnapshot('early win');
});

it('should run multiple rounds with the winner on the last only', () => {
  const {rounds} = getResults(balanced, candidates);
  const firstRound = rounds[0];
  const secondRound = rounds[1];
  const lastRound = rounds[rounds.length - 1];

  expect(rounds.length).toEqual(3);
  expect(firstRound.winners.length).toEqual(0);
  expect(secondRound.winners.length).toEqual(0);
  expect(lastRound.winners[0]).toEqual('D');
  expect(rounds).toMatchSnapshot('balanced win');
});

it('should provide full segment lists for each candidate', () => {
  const {rounds} = getResults(balanced, candidates);
  const firstRound = rounds[0];
  const secondRound = rounds[1];
  const lastRound = rounds[rounds.length - 1];

  expect(rounds.length).toEqual(3);
  expect(firstRound.winners.length).toEqual(0);
  expect(secondRound.winners.length).toEqual(0);
  expect(lastRound.winners[0]).toEqual('D');
  expect(rounds).toMatchSnapshot('balanced win');
});
