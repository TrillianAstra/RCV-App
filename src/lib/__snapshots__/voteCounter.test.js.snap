// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`balanced win 1`] = `
Array [
  Object {
    "loser": "A",
    "previousLosers": Array [],
    "segments": Object {
      "A": Map {
        "A" => 1,
        "B" => 0,
        "C" => 0,
        "D" => 0,
        "E" => 0,
      },
      "B": Map {
        "B" => 1,
        "A" => 0,
        "C" => 0,
        "D" => 0,
        "E" => 0,
      },
      "C": Map {
        "C" => 1,
        "A" => 0,
        "B" => 0,
        "D" => 0,
        "E" => 0,
      },
      "D": Map {
        "D" => 2,
        "A" => 0,
        "B" => 0,
        "C" => 0,
        "E" => 0,
      },
      "E": Map {
        "E" => 2,
        "A" => 0,
        "B" => 0,
        "C" => 0,
        "D" => 0,
      },
    },
    "totals": Object {
      "A": 1,
      "B": 1,
      "C": 1,
      "D": 2,
      "E": 2,
    },
    "validVoteCount": 7,
    "winner": undefined,
  },
  Object {
    "loser": "B",
    "previousLosers": Array [
      "A",
    ],
    "segments": Object {
      "A": Map {
        "A" => 0,
        "B" => 0,
        "C" => 0,
        "D" => 0,
        "E" => 0,
      },
      "B": Map {
        "B" => 1,
        "A" => 0,
        "C" => 0,
        "D" => 0,
        "E" => 0,
      },
      "C": Map {
        "C" => 1,
        "A" => 0,
        "B" => 0,
        "D" => 0,
        "E" => 0,
      },
      "D": Map {
        "D" => 2,
        "A" => 1,
        "B" => 0,
        "C" => 0,
        "E" => 0,
      },
      "E": Map {
        "E" => 2,
        "A" => 0,
        "B" => 0,
        "C" => 0,
        "D" => 0,
      },
    },
    "totals": Object {
      "A": 0,
      "B": 1,
      "C": 1,
      "D": 3,
      "E": 2,
    },
    "validVoteCount": 7,
    "winner": undefined,
  },
  Object {
    "loser": "C",
    "previousLosers": Array [
      "A",
      "B",
    ],
    "segments": Object {
      "A": Map {
        "A" => 0,
        "B" => 0,
        "C" => 0,
        "D" => 0,
        "E" => 0,
      },
      "B": Map {
        "B" => 0,
        "A" => 0,
        "C" => 0,
        "D" => 0,
        "E" => 0,
      },
      "C": Map {
        "C" => 1,
        "A" => 0,
        "B" => 0,
        "D" => 0,
        "E" => 0,
      },
      "D": Map {
        "D" => 2,
        "A" => 1,
        "B" => 1,
        "C" => 0,
        "E" => 0,
      },
      "E": Map {
        "E" => 2,
        "A" => 0,
        "B" => 0,
        "C" => 0,
        "D" => 0,
      },
    },
    "totals": Object {
      "A": 0,
      "B": 0,
      "C": 1,
      "D": 4,
      "E": 2,
    },
    "validVoteCount": 7,
    "winner": "D",
  },
]
`;

exports[`balanced win 2`] = `
Array [
  Object {
    "loser": "A",
    "previousLosers": Array [],
    "segments": Object {
      "A": Map {
        "A" => 1,
        "B" => 0,
        "C" => 0,
        "D" => 0,
        "E" => 0,
      },
      "B": Map {
        "B" => 1,
        "A" => 0,
        "C" => 0,
        "D" => 0,
        "E" => 0,
      },
      "C": Map {
        "C" => 1,
        "A" => 0,
        "B" => 0,
        "D" => 0,
        "E" => 0,
      },
      "D": Map {
        "D" => 2,
        "A" => 0,
        "B" => 0,
        "C" => 0,
        "E" => 0,
      },
      "E": Map {
        "E" => 2,
        "A" => 0,
        "B" => 0,
        "C" => 0,
        "D" => 0,
      },
    },
    "totals": Object {
      "A": 1,
      "B": 1,
      "C": 1,
      "D": 2,
      "E": 2,
    },
    "validVoteCount": 7,
    "winner": undefined,
  },
  Object {
    "loser": "B",
    "previousLosers": Array [
      "A",
    ],
    "segments": Object {
      "A": Map {
        "A" => 0,
        "B" => 0,
        "C" => 0,
        "D" => 0,
        "E" => 0,
      },
      "B": Map {
        "B" => 1,
        "A" => 0,
        "C" => 0,
        "D" => 0,
        "E" => 0,
      },
      "C": Map {
        "C" => 1,
        "A" => 0,
        "B" => 0,
        "D" => 0,
        "E" => 0,
      },
      "D": Map {
        "D" => 2,
        "A" => 1,
        "B" => 0,
        "C" => 0,
        "E" => 0,
      },
      "E": Map {
        "E" => 2,
        "A" => 0,
        "B" => 0,
        "C" => 0,
        "D" => 0,
      },
    },
    "totals": Object {
      "A": 0,
      "B": 1,
      "C": 1,
      "D": 3,
      "E": 2,
    },
    "validVoteCount": 7,
    "winner": undefined,
  },
  Object {
    "loser": "C",
    "previousLosers": Array [
      "A",
      "B",
    ],
    "segments": Object {
      "A": Map {
        "A" => 0,
        "B" => 0,
        "C" => 0,
        "D" => 0,
        "E" => 0,
      },
      "B": Map {
        "B" => 0,
        "A" => 0,
        "C" => 0,
        "D" => 0,
        "E" => 0,
      },
      "C": Map {
        "C" => 1,
        "A" => 0,
        "B" => 0,
        "D" => 0,
        "E" => 0,
      },
      "D": Map {
        "D" => 2,
        "A" => 1,
        "B" => 1,
        "C" => 0,
        "E" => 0,
      },
      "E": Map {
        "E" => 2,
        "A" => 0,
        "B" => 0,
        "C" => 0,
        "D" => 0,
      },
    },
    "totals": Object {
      "A": 0,
      "B": 0,
      "C": 1,
      "D": 4,
      "E": 2,
    },
    "validVoteCount": 7,
    "winner": "D",
  },
]
`;

exports[`early win 1`] = `
Array [
  Object {
    "loser": "B",
    "previousLosers": Array [],
    "segments": Object {
      "A": Map {
        "A" => 1,
        "B" => 0,
        "C" => 0,
        "D" => 0,
        "E" => 0,
      },
      "B": Map {
        "B" => 0,
        "A" => 0,
        "C" => 0,
        "D" => 0,
        "E" => 0,
      },
      "C": Map {
        "C" => 1,
        "A" => 0,
        "B" => 0,
        "D" => 0,
        "E" => 0,
      },
      "D": Map {
        "D" => 4,
        "A" => 0,
        "B" => 0,
        "C" => 0,
        "E" => 0,
      },
      "E": Map {
        "E" => 1,
        "A" => 0,
        "B" => 0,
        "C" => 0,
        "D" => 0,
      },
    },
    "totals": Object {
      "A": 1,
      "B": 0,
      "C": 1,
      "D": 4,
      "E": 1,
    },
    "validVoteCount": 7,
    "winner": "D",
  },
]
`;
